File: packages/edge/internal/server/access.go
package server

import (
	"encoding/json"
	"log"
	"net/http"
)

type rankingDTO struct {
	ID       string `json:"id"`
	Title    string `json:"title"`
	ImageURL string `json:"imageUrl"`
	Rank     int    `json:"rank"`
	Start    string `json:"start"`
}

type rankingsResponse struct {
	Rankings  []rankingDTO `json:"rankings"`
	FetchedAt *string      `json:"fetchedAt"`
}

func (s *Server) handleGetRankings(w http.ResponseWriter, r *http.Request) {
	rankings, fetchedAt, err := s.rankings.ListReady(r.Context())
	if err != nil {
		log.Printf("rankings list: %v", err)
		http.Error(w, `{"error":"internal error"}`, http.StatusInternalServerError)
		return
	}

	res := rankingsResponse{
		Rankings: make([]rankingDTO, 0, len(rankings)),
	}
	for _, rk := range rankings {
		res.Rankings = append(res.Rankings, rankingDTO{
			ID:       rk.ID,
			Title:    rk.Title,
			ImageURL: joinMediaURL(s.cfg.MediaURLPrefix, rk.ImageURL),
			Rank:     rk.Rank,
			Start:    rk.Start.UTC().Format("2006-01-02T15:04:05.000Z"),
		})
	}
	if fetchedAt != nil {
		s := fetchedAt.UTC().Format("2006-01-02T15:04:05.000Z")
		res.FetchedAt = &s
	}

	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(res)
}

File: packages/signage/src/api/access.ts
export interface RankingItem {
  id: string;
  title: string;
  imageUrl: string;
  rank: number;
  start: string;
}

export interface RankingsData {
  rankings: RankingItem[];
  fetchedAt: string | null;
}

interface RankingsResponse {
  rankings: RankingItem[];
  fetchedAt: string | null;
}

const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8080/api";

export async function fetchRankings(): Promise<RankingsData> {
  const res = await fetch(`${API_BASE_URL}/signage/rankings`);
  if (!res.ok) {
    throw new Error(`Failed to fetch rankings: ${res.status}`);
  }
  const data: RankingsResponse = await res.json();
  return { rankings: data.rankings, fetchedAt: data.fetchedAt };
}
